### PR TITLE
[Feature][starter][flink] Support transform-v2 for flink 

### PR DESCRIPTION
**Pull request by john8628**
## Purpose of this pull request

https://github.com/apache/incubator-seatunnel/issues/3267

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)

### CODE REVIEW

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformExecuteProcessor.java [hunk 1]
```diff
@@ -34,6 +34,7 @@
 import java.util.List;
 import java.util.stream.Collectors;
 
+@Deprecated
 public class TransformExecuteProcessor extends AbstractPluginExecuteProcessor<FlinkStreamTransform> {
```
Comment (hailin0) [inline seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformExecuteProcessor.java]: remove this file?

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformV2ExecuteProcessor.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    @Override
+    public List<DataStream<Row>> execute(List<DataStream<Row>> upstreamDataStreams) throws TaskExecuteException {
+        if (plugins.isEmpty()) {
+            return upstreamDataStreams;
+        }
+        DataStream<Row> input = upstreamDataStreams.get(0);
+        List<DataStream<Row>> result = new ArrayList<>();
+        for (int i = 0; i < plugins.size(); i++) {
+            try {
+                SeaTunnelTransform<Row> transform = plugins.get(i);
```
Comment (hailin0) [inline seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformV2ExecuteProcessor.java]:
```suggestion
                SeaTunnelTransform<SeaTunnelRow> transform = plugins.get(i);
```

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformV2ExecuteProcessor.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                result.add(input);
+            } catch (Exception e) {
+                throw new TaskExecuteException(
+                    String.format("SeaTunnel transform task: %s execute error", plugins.get(i).getPluginName()), e);
+            }
+        }
+        return result;
+    }
+
+    protected DataStream<Row> flinkTransformV2(SeaTunnelTransform transform, DataStream<Row> stream) {
```
Comment (hailin0) [inline seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformV2ExecuteProcessor.java]:
```suggestion
    protected DataStream<Row> flinkTransform(SeaTunnelTransform transform, DataStream<Row> stream) {
```

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformV2ExecuteProcessor.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+    private static final String PLUGIN_TYPE = "transform";
+
+    protected TransformV2ExecuteProcessor(List<URL> jarPaths, List<? extends Config> pluginConfigs, JobContext jobContext) {
+        super(jarPaths, pluginConfigs, jobContext);
+    }
+
+    @Override
+    protected List<SeaTunnelTransform> initializePlugins(List<URL> jarPaths, List<? extends Config> pluginConfigs, JobContext jobContext) {
+        SeaTunnelTransformPluginDiscovery transformV2PluginDiscovery = new SeaTunnelTransformPluginDiscovery();
```
Comment (hailin0) [inline seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformV2ExecuteProcessor.java]:
```suggestion
        SeaTunnelTransformPluginDiscovery transformPluginDiscovery = new SeaTunnelTransformPluginDiscovery();
```

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformV2ExecuteProcessor.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        }
+        return result;
+    }
+
+    protected DataStream<Row> flinkTransformV2(SeaTunnelTransform transform, DataStream<Row> stream) {
+        SeaTunnelDataType seaTunnelDataType = TypeConverterUtils.convert(stream.getType());
+        transform.setTypeInfo(seaTunnelDataType);
+        TypeInformation rowTypeInfo = TypeConverterUtils.convert(transform.getProducedType());
+        FlinkRowConverter transformInputRowConverter = new FlinkRowConverter(seaTunnelDataType);
+        FlinkRowConverter transformOutRowConverter = new FlinkRowConverter(transform.getProducedType());
```
Comment (hailin0) [inline seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformV2ExecuteProcessor.java]:
```suggestion
        FlinkRowConverter transformOutputRowConverter = new FlinkRowConverter(transform.getProducedType());
```

File: seatunnel-engine/seatunnel-engine-client/src/test/resources/batch_fakesource_to_file.conf [hunk 6]
```diff
@@ -41,6 +41,12 @@ source {
 }
 
 transform {
+  Copy {
+    source_table_name = "fake"
+    result_table_name = "fake1"
+    src_field = "name"
+    dest_field = "name1"
+  }
 }
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-client/src/test/resources/batch_fakesource_to_file.conf]: revert

File: seatunnel-examples/seatunnel-engine-examples/src/main/resources/examples/fake_to_console.conf [hunk 7]
```diff
@@ -52,6 +52,12 @@ source {
 }
 
 transform {
+  Copy {
+    source_table_name = "fake"
+    result_table_name = "fake1"
+    src_field = "name"
+    dest_field = "name1"
+  }
```
Comment (hailin0) [inline seatunnel-examples/seatunnel-engine-examples/src/main/resources/examples/fake_to_console.conf]: revert

File: seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml [hunk 8]
```diff
@@ -96,6 +96,13 @@
             <scope>${flink.scope}</scope>
         </dependency>
 
+        <dependency>
+            <groupId>org.apache.seatunnel</groupId>
+            <artifactId>seatunnel-transforms-v2</artifactId>
+            <version>${project.version}</version>
+            <scope>${flink.scope}</scope>
+        </dependency>
+
```
Comment (hailin0) [inline seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml]: revert

example module only run as your local env

File: seatunnel-examples/seatunnel-flink-connector-v2-example/src/main/java/org/apache/seatunnel/example/flink/v2/SeaTunnelApiExample.java [hunk 9]
```diff
@@ -31,7 +31,7 @@
 public class SeaTunnelApiExample {
 
     public static void main(String[] args) throws FileNotFoundException, URISyntaxException, CommandException {
-        String configurePath = args.length > 0 ?  args[0] : "/examples/fake_to_console.conf";
+        String configurePath = args.length > 0 ?  args[0] : "/examples/split_transform.conf";
```
Comment (hailin0) [inline seatunnel-examples/seatunnel-flink-connector-v2-example/src/main/java/org/apache/seatunnel/example/flink/v2/SeaTunnelApiExample.java]: revert

**Review by hailin0 — COMMENTED** [review]
transformv-v2 doesn't support sql, you also need to fix older versions of e2e testcase

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/AbstractPluginExecuteProcessor.java [hunk 10]
```diff
@@ -50,15 +50,15 @@
     protected AbstractPluginExecuteProcessor(List<URL> jarPaths, List<? extends Config> pluginConfigs, JobContext jobContext) {
         this.pluginConfigs = pluginConfigs;
         this.jobContext = jobContext;
-        this.plugins = initializePlugins(jarPaths, pluginConfigs);
+        this.plugins = initializePlugins(jarPaths, pluginConfigs, jobContext);
```
Comment (FWLamb) [inline seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/AbstractPluginExecuteProcessor.java]: I think we can use jobContext directly in initializePlugins without adding parameters.

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformExecuteProcessor.java (hunk 1 above)
Comment (john8628) [reply seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformExecuteProcessor.java]: OK

File: seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml (hunk 8 above)
Comment (john8628) [reply seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml]: ok ,thanks

File: seatunnel-engine/seatunnel-engine-client/src/test/resources/batch_fakesource_to_file.conf (hunk 6 above)
Comment (john8628) [reply seatunnel-engine/seatunnel-engine-client/src/test/resources/batch_fakesource_to_file.conf]: OK

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/AbstractPluginExecuteProcessor.java (hunk 10 above)
Comment (john8628) [reply seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/AbstractPluginExecuteProcessor.java]: But initializePlugins's config will only include the specified(transform,souce,sink) config;

Comment (john8628) [reply seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/AbstractPluginExecuteProcessor.java]: ok, got it ,thanks

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-iotdb-flink-e2e/src/test/resources/iotdb/iotdb_source_to_sink.conf [hunk 11]
```diff
@@ -54,9 +54,7 @@ source {
 }
 
 transform {
-    sql {
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-iotdb-flink-e2e/src/test/resources/iotdb/iotdb_source_to_sink.conf]: You also need to fix the logic in the test case, not just remove

Comment (john8628) [reply seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-iotdb-flink-e2e/src/test/resources/iotdb/iotdb_source_to_sink.conf]: OK ,thx ;

**Comment by hailin0** [comment]
https://github.com/apache/incubator-seatunnel/actions/runs/3462689789/jobs/5781983015
image

**Review by FWLamb — DISMISSED** [review]

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/resources/assertion/fakesource_to_assert.conf [hunk 12]
```diff
@@ -41,16 +41,19 @@ source {
...
+      fields = ["name", "age"]
+    }
 
   # If you would like to get more information about how to configure seatunnel and see full list of transform plugins,
   # please go to https://seatunnel.apache.org/docs/flink/configuration/transform-plugins/Sql
 }
 
 sink {
   Assert {
+  source_table_name = "fake1"
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/resources/assertion/fakesource_to_assert.conf]:
```suggestion
    source_table_name = "fake1"
```

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-datahub-flink-e2e/src/test/resources/datahub/fakesource_to_datahub.conf [hunk 13]
```diff
@@ -42,16 +42,19 @@ source {
...
 
 transform {
-    sql {
-      sql = "select name,age from fake"
-    }
+     Filter {
+        source_table_name = "fake"
+        result_table_name = "fake1"
+        fields = ["name", "age"]
+     }
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-datahub-flink-e2e/src/test/resources/datahub/fakesource_to_datahub.conf]: remove

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-datahub-flink-e2e/src/test/resources/datahub/fakesource_to_datahub.conf [hunk 14]
```diff
@@ -42,16 +42,19 @@ source {
...
 
   # If you would like to get more information about how to configure seatunnel and see full list of transform plugins,
   # please go to https://seatunnel.apache.org/docs/transform/sql
 }
 
 sink {
   DataHub {
+      source_table_name = "fake1"
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-datahub-flink-e2e/src/test/resources/datahub/fakesource_to_datahub.conf]: remove

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-fake-flink-e2e/src/test/resources/fake/fakesource_to_console.conf [hunk 15]
```diff
@@ -42,10 +42,13 @@ source {
...
 transform {
-    sql {
-      sql = "select name,age from fake"
+    Filter {
+            source_table_name = "fake"
+            result_table_name = "fake1"
+            fields = ["name", "age"]
     }
 
+
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-fake-flink-e2e/src/test/resources/fake/fakesource_to_console.conf]: remove

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-iotdb-flink-e2e/src/test/resources/iotdb/iotdb_source_to_sink.conf [hunk 16]
```diff
@@ -54,13 +54,28 @@ source {
...
+  }
+Replace {
+    source_table_name = "fake1"
+    result_table_name = "fake2"
+    replace_field = "device_name"
+    pattern = "root.source_group.device_b"
+    replacement = "root.sink_group.device_b"
+    is_regex = false
+  }
+
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-iotdb-flink-e2e/src/test/resources/iotdb/iotdb_source_to_sink.conf]:
```suggestion
  Replace {
    source_table_name = "fake"
    result_table_name = "fake1"
    replace_field = "device_name"
    pattern = "root.source_group"
    replacement = "root.sink_group"
    is_regex = false
  }

```

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-iotdb-flink-e2e/src/test/resources/iotdb/iotdb_source_to_sink.conf [hunk 17]
```diff
@@ -54,13 +54,28 @@ source {
...
 }
 
 sink {
     IoTDB {
+        source_table_name = "fake2"
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-iotdb-flink-e2e/src/test/resources/iotdb/iotdb_source_to_sink.conf]:
```suggestion
        source_table_name = "fake1"
```

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/resources/jdbc/fakesource_to_jdbc.conf [hunk 18]
```diff
@@ -42,17 +42,18 @@ source {
 }
 
 transform {
-    sql {
-      sql = "select name from fake"
+    Filter {
+            source_table_name = "fake"
+            result_table_name = "fake1"
+            fields = ["name"]
     }
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/resources/jdbc/fakesource_to_jdbc.conf]: remove

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/resources/jdbc/fakesource_to_jdbc.conf [hunk 19]
```diff
@@ -42,17 +42,18 @@ source {
...
-
   # If you would like to get more information about how to configure seatunnel and see full list of transform plugins,
   # please go to https://seatunnel.apache.org/docs/transform/sql
 }
 
 sink {
   Jdbc {
-      source_table_name = fake
+      source_table_name = fake1
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/resources/jdbc/fakesource_to_jdbc.conf]: remove

File: seatunnel-e2e/seatunnel-transforms-v2-e2e/src/test/java/org/apache/seatunnel/e2e/transform/TestSuiteBase.java [hunk 20]
```diff
@@ -37,7 +37,7 @@
     TestCaseInvocationContextProvider.class
 })
 @TestInstance(TestInstance.Lifecycle.PER_CLASS)
-@DisabledOnContainer(value = {}, type = {EngineType.FLINK, EngineType.SPARK}, disabledReason = "TODO: Transform v2 translation to spark/flink isn't completed")
+@DisabledOnContainer(value = {}, type = {EngineType.SPARK}, disabledReason = "TODO: Transform v2 translation to spark/flink isn't completed")
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-transforms-v2-e2e/src/test/java/org/apache/seatunnel/e2e/transform/TestSuiteBase.java]:
```suggestion
@DisabledOnContainer(value = {}, type = {EngineType.SPARK}, disabledReason = "TODO: Transform v2 translation to spark isn't completed")
```

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java [hunk 21]
```diff
@@ -58,8 +58,8 @@ public class FlinkExecution implements TaskExecution {
 
     private final FlinkEnvironment flinkEnvironment;
     private final PluginExecuteProcessor sourcePluginExecuteProcessor;
-    private final PluginExecuteProcessor transformPluginExecuteProcessor;
```
Comment (hailin0) [inline seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java]: revert this file change

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformV2ExecuteProcessor.java [hunk 22]
```diff
@@ -34,26 +40,28 @@
...
     private static final String PLUGIN_TYPE = "transform";
 
-    protected TransformExecuteProcessor(List<URL> jarPaths, List<? extends Config> pluginConfigs, JobContext jobContext) {
+    protected TransformV2ExecuteProcessor(List<URL> jarPaths, List<? extends Config> pluginConfigs, JobContext jobContext) {
         super(jarPaths, pluginConfigs, jobContext);
     }
 
     @Override
-    protected List<FlinkStreamTransform> initializePlugins(List<URL> jarPaths, List<? extends Config> pluginConfigs) {
-        SeaTunnelFlinkTransformPluginDiscovery transformPluginDiscovery = new SeaTunnelFlinkTransformPluginDiscovery();
```
Comment (hailin0) [inline seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformV2ExecuteProcessor.java:47]: remove this file `org.apache.seatunnel.plugin.discovery.seatunnel.SeaTunnelFlinkTransformPluginDiscovery`

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformV2ExecuteProcessor.java [hunk 23]
```diff
@@ -82,4 +89,23 @@ public List<DataStream<Row>> execute(List<DataStream<Row>> upstreamDataStreams)
...
+        FlinkRowConverter transformOutRowConverter = new FlinkRowConverter(transform.getProducedType());
+        DataStream<Row> output =  stream.map(new MapFunction<Row, Row>() {
+            @Override
+            public Row map(Row value) throws Exception {
+                SeaTunnelRow seaTunnelRow = transformInputRowConverter.reconvert(value);
+                SeaTunnelRow dataRow = (SeaTunnelRow) transform.map(seaTunnelRow);
+                Row copy = transformOutRowConverter.convert(dataRow);
+                return copy;
+            }
+        }).returns(rowTypeInfo);
```
Comment (hailin0) [inline seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformV2ExecuteProcessor.java]:
```suggestion
        DataStream<Row> output =  stream.map(new MapFunction<Row, Row>() {
            @Override
            public Row map(Row value) throws Exception {
                SeaTunnelRow seaTunnelRow = transformInputRowConverter.reconvert(value);
                SeaTunnelRow dataRow = (SeaTunnelRow) transform.map(seaTunnelRow);
                Row copy = transformOutRowConverter.convert(dataRow);
                return copy;
            }
        },
        rowTypeInfo);
```

**Comment by hailin0** [comment]
please fix ci error

**Comment by FWLamb** [comment]
LGTM

**Review by FWLamb — APPROVED** [review]

**Comment by EricJoy2048** [comment]
@hailin0  PTAL.

**Review by hailin0 — APPROVED** [review]
LGTM

**Review by TaoZex — APPROVED** [review]
+1

**Review by EricJoy2048 — APPROVED** [review]

**Pull request merged**